# Titanic Kaggle Project

In [1]:
#importing all the required libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib notebook

In [2]:
#importing the dataset  train.csv

df = pd.read_csv('train.csv')
rt= df.copy()
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df.shape

(891, 12)

It has 891 observations and 12 features

survived is the target label

In [4]:
df.isna().values.any()

True

In [5]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [7]:
df.PassengerId.isna().any()

False

In [8]:
df.SibSp.unique()

array([1, 0, 3, 4, 2, 5, 8])

In [9]:
df.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [10]:
df.Parch.unique()

array([0, 1, 2, 5, 3, 4, 6])

In [11]:
df.Ticket.unique().size

681

In [12]:
df.Pclass.unique()

array([3, 1, 2])

In [13]:
df.Fare.unique().size

248

In [14]:
df[df.Cabin.isna() == True].shape[0]

687

In [15]:
df[df.Survived == 1].shape

(342, 12)

In [16]:
df[df.Survived == 0].shape

(549, 12)

# ---------------------------------------------------------------------------------------------------------------

The 'Cabin' feature has 687 null values so we are not going to consider it as one of  the final feature 

And also Name, Ticket, Cabin are also not considered for the final feature reasons--
- Name has nothing to do with it as long as we have the passenger Id.
- Ticket has 687 unique values and it does'nt give any extra insight about the passenger
- Cabin feature has 687 null values which will make the model unstable and the target label cannot depend on it.

So we look at the remaining features Passenger Id, Pclass, Sex, Age, Fare, SibSp, Embarked.

In [17]:
#droping all the colums mentioned above

df.drop(['Cabin','Name','Ticket'],inplace=True,axis=1)

## 0.Filling all the Null values

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          714 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Fare         891 non-null    float64
 8   Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 62.8+ KB


# -----------------------------------------------------------------------------------------------------------------

we have null values for Age,Embarked.

we are not going to drop the null values as it reduces the quality of the data
i.e over more than 100 rows will be dropped so that reduces the weightness of the data.

Instead we will be using the basic mean filling method for the Age column.


# -----------------------------------------------------------------------------------------------------------------

In [19]:
m_fill = df.Age.mean()

df.Age.replace(np.nan ,m_fill,inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Fare         891 non-null    float64
 8   Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 62.8+ KB


  #Now for the Embarked feature we are going to use the Mode for filling all the NaN values. As Embarked is a categorical value.

In [20]:
median_fill = df.Embarked.mode()[0]
df.Embarked.replace(np.nan ,median_fill ,inplace=True)
df.info()
rt = df.copy()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Fare         891 non-null    float64
 8   Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 62.8+ KB


## 1.Covert Objects to numerical data

In [163]:
# sex_dict = {'male':1,'female':0}
# embarked_dict = dict(zip( rt.Embarked.unique(),np.arange(0,len(rt.Embarked.unique()))))
# https://mario-vishal.github.io/Todo/

In [24]:
df =pd.get_dummies(df)

In [25]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Fare  \
0              1         0       3  22.000000      1      0   7.2500   
1              2         1       1  38.000000      1      0  71.2833   
2              3         1       3  26.000000      0      0   7.9250   
3              4         1       1  35.000000      1      0  53.1000   
4              5         0       3  35.000000      0      0   8.0500   
..           ...       ...     ...        ...    ...    ...      ...   
886          887         0       2  27.000000      0      0  13.0000   
887          888         1       1  19.000000      0      0  30.0000   
888          889         0       3  29.699118      1      2  23.4500   
889          890         1       1  26.000000      0      0  30.0000   
890          891         0       3  32.000000      0      0   7.7500   

     Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S  
0             0         1           0           0           1  
1             1         0           1           0           0  
2             1         0           0           0           1  
3             1         0           0           0           1  
4             0         1           0           0           1  
..          ...       ...         ...         ...         ...  
886           0         1           0           0           1  
887           1         0           0           0           1  
888           1         0           0           0           1  
889           0         1           1           0           0  
890           0         1           0           1           0  

[891 rows x 12 columns]

## 2.Split the Data into X,y and train_set,test_sets

Before splitting the data into features and target labels first lets remove the passenger Id from the data

In [26]:
df.drop(['PassengerId'],inplace=True,axis=1)

In [27]:
df.head()

Survived  Pclass   Age  SibSp  Parch     Fare  Sex_female  Sex_male  \
0         0       3  22.0      1      0   7.2500           0         1   
1         1       1  38.0      1      0  71.2833           1         0   
2         1       3  26.0      0      0   7.9250           1         0   
3         1       1  35.0      1      0  53.1000           1         0   
4         0       3  35.0      0      0   8.0500           0         1   

   Embarked_C  Embarked_Q  Embarked_S  
0           0           0           1  
1           1           0           0  
2           0           0           1  
3           0           0           1  
4           0           0           1

In [29]:
X = df.iloc[:,1:]
y = df.iloc[:,0]
y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [30]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train ,y_test,= train_test_split(X,y,random_state=0)

#70% train set and 30% test set

In [31]:
X_train

Pclass        Age  SibSp  Parch     Fare  Sex_female  Sex_male  \
105       3  28.000000      0      0   7.8958           0         1   
68        3  17.000000      4      2   7.9250           1         0   
253       3  30.000000      1      0  16.1000           0         1   
320       3  22.000000      0      0   7.2500           0         1   
706       2  45.000000      0      0  13.5000           1         0   
..      ...        ...    ...    ...      ...         ...       ...   
835       1  39.000000      1      1  83.1583           1         0   
192       3  19.000000      1      0   7.8542           1         0   
629       3  29.699118      0      0   7.7333           0         1   
559       3  36.000000      1      0  17.4000           1         0   
684       2  60.000000      1      1  39.0000           0         1   

     Embarked_C  Embarked_Q  Embarked_S  
105           0           0           1  
68            0           0           1  
253           0           0           1  
320           0           0           1  
706           0           0           1  
..          ...         ...         ...  
835           1           0           0  
192           0           0           1  
629           0           1           0  
559           0           0           1  
684           0           0           1  

[668 rows x 10 columns]

In [32]:
y_train

105    0
68     1
253    0
320    0
706    1
      ..
835    1
192    1
629    0
559    1
684    0
Name: Survived, Length: 668, dtype: int64

## 3.Building the model

Let's now import Logistic regression from linear model and try to fit the data to the model and observe it's accuracy!

### Logistic Regression

In [112]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import r2_score

lr = LogisticRegression(max_iter=550)

#fitting the data to the model

lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=550,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [34]:
lr.score(X_train,y_train)

0.8038922155688623

In [121]:
r2_score(y_test,lr.predict(X_test))

0.1214457005823909

### Decision Tree Classifier

In [120]:
from sklearn.tree import DecisionTreeClassifier

dtr = DecisionTreeClassifier(max_depth=12)

dtr.fit(X_train,y_train)
y_pre_dtr = dtr.predict(X_test)
r2_score(y_test,y_pre_dtr)

0.17874272010962633

In [43]:
dict(zip(dtr.feature_importances_*100,X_train.columns))

{10.728955232313218: 'Pclass',
 20.481980191846656: 'Age',
 7.371416891035547: 'SibSp',
 2.75992451420093: 'Parch',
 18.508038965199187: 'Fare',
 0.0: 'Embarked_Q',
 38.96190271736805: 'Sex_male',
 0.3509065529363399: 'Embarked_C',
 0.8368749351000659: 'Embarked_S'}

## SVC

In [95]:
from sklearn.svm import SVC

svc = SVC(C=20,kernel='rbf')
svc.fit(X_train,y_train)

SVC(C=20, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [96]:
svc.score(X_train,y_train)

0.8023952095808383

In [118]:
y_pred_test = svc.predict(X_test)
r2_score(y_test,y_pred_test)

0.1978417266187048

## Random forest

In [154]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier()
clf.fit(X_train,y_train)
clf.score(X_train,y_train)

0.9805389221556886

In [155]:
clf.score(X_test,y_test)

0.8430493273542601

In [157]:
y_pre_clf = clf.predict(X_test)

In [158]:
r2_score(y_test,y_pre_clf)

0.331534772182254

## Importing the test set which do not have target labels

In [139]:
df_test = pd.read_csv('test.csv')
df_p = df_test.copy()
df_test.drop(['PassengerId','Cabin','Name','Ticket'],axis=1,inplace=True)
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    418 non-null    int64  
 1   Sex       418 non-null    object 
 2   Age       332 non-null    float64
 3   SibSp     418 non-null    int64  
 4   Parch     418 non-null    int64  
 5   Fare      417 non-null    float64
 6   Embarked  418 non-null    object 
dtypes: float64(2), int64(3), object(2)
memory usage: 23.0+ KB


In [146]:
m_fill = df_test.Age.mean()

df_test.Age.replace(np.nan ,m_fill,inplace=True)

df_test.Fare.replace(np.nan,df_test.Fare.median(),inplace=True)
df_test.info()
df_test = pd.get_dummies(df_test)
df_test

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    418 non-null    int64  
 1   Sex       418 non-null    object 
 2   Age       418 non-null    float64
 3   SibSp     418 non-null    int64  
 4   Parch     418 non-null    int64  
 5   Fare      418 non-null    float64
 6   Embarked  418 non-null    object 
dtypes: float64(2), int64(3), object(2)
memory usage: 23.0+ KB


Pclass       Age  SibSp  Parch      Fare  Sex_female  Sex_male  \
0         3  34.50000      0      0    7.8292           0         1   
1         3  47.00000      1      0    7.0000           1         0   
2         2  62.00000      0      0    9.6875           0         1   
3         3  27.00000      0      0    8.6625           0         1   
4         3  22.00000      1      1   12.2875           1         0   
..      ...       ...    ...    ...       ...         ...       ...   
413       3  30.27259      0      0    8.0500           0         1   
414       1  39.00000      0      0  108.9000           1         0   
415       3  38.50000      0      0    7.2500           0         1   
416       3  30.27259      0      0    8.0500           0         1   
417       3  30.27259      1      1   22.3583           0         1   

     Embarked_C  Embarked_Q  Embarked_S  
0             0           1           0  
1             0           0           1  
2             0           1           0  
3             0           0           1  
4             0           0           1  
..          ...         ...         ...  
413           0           0           1  
414           1           0           0  
415           0           0           1  
416           0           0           1  
417           1           0           0  

[418 rows x 10 columns]

### I have tried with both Decision tree and Logistic regression but Decision tree has seemed to work pretty well

In [159]:
y_pred = clf.predict(df_test)

In [160]:
df_p['Survived'] = y_pred

In [161]:
df_results = df_p[['PassengerId','Survived']]
df_results

PassengerId  Survived
0            892         0
1            893         0
2            894         0
3            895         1
4            896         0
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         1

[418 rows x 2 columns]

### Writing our results to .csv file

In [162]:
df_results.to_csv('titanic_results_clf.csv',index=False)

## Result -

I have got an accuracy of 75% for logistic regression 
and 
for Decision tree an accuracy of 76%.